pkg/skaffold/instrumentation/trace.go

```diff
@@ -0,0 +1,85 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package instrumentation
+
+import (
+	"context"
+	"os"
+	"sync"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+)
+
+var traceEnabled bool
+var traceInitOnce sync.Once
+
+var tracerProvider trace.TracerProvider
+var tracerShutdown func(context.Context) error = func(context.Context) error { return nil }
+var tracerInitErr error
+
+// InitTraceFromEnvVar initializes the singleton skaffold tracer from the SKAFFOLD_TRACE env variable.
+// The code here is a wrapper around the opentelemetry(otel) trace libs for usability
+// When SKAFFOLD_TRACE is set, this will setup the proper tracer provider (& exporter),
+// configures otel to use this tracer provider and saves the  tracer provider shutdown function
+// to be used globally so that it can be run before skaffold exits.
+func InitTraceFromEnvVar(opts ...TraceExporterOption) (trace.TracerProvider, func(context.Context) error, error) {
+	traceInitOnce.Do(func() {
+		_, skaffTrace := os.LookupEnv("SKAFFOLD_TRACE")
+		_, skaffExportToStdout := os.LookupEnv("SKAFFOLD_EXPORT_TO_STDOUT")
+		if skaffTrace || skaffExportToStdout {
+			traceEnabled = true
+		}
+		if traceEnabled {
+			tp, shutdown, err := initTraceExporter(opts...)
+			otel.SetTracerProvider(tp)
+			tracerProvider = tp
+			tracerShutdown = shutdown
+			tracerInitErr = err
+		}
+	})
+	return tracerProvider, tracerShutdown, tracerInitErr
+}
+
+// TracerShutdown is a function used to flush all current running spans and make sure they are exported.  This should be called once
+// at the end of a skaffold run to properly shutdown and export all spans for the singleton.
+func TracerShutdown(ctx context.Context) error {
+	traceInitOnce = sync.Once{}
+	return tracerShutdown(ctx)
+}
+
+// StartTrace uses the otel trace provider to export timing spans (with optional attributes) to the chosen exporter
+//  via the value set in SKAFFOLD_TRACE.  Tracing is done via metadata stored in a context.Context, this means that to properly
+// get parent/child traces a consistent context should be used.  The returned context will have the span metadata added so this context should
+// be used for subsequent calls in skaffold.  The returned function should be called to end the trace span, oftentimes this can be done with
+// the form:  _, endTrace = StartTrace...; defer endTrace()
+func StartTrace(ctx context.Context, pkg, name string, attributes ...map[string]string) (context.Context, func(options ...trace.SpanOption)) {
```

Comment:
You could skip `pkg` and use Go's [`runtime.Caller()`](https://golang.org/pkg/runtime/#Caller) to get the file and location.  If there are any restrictions on the `name` you should mention it above.

Reply:
Removed `pkg` param and added `runtime.Caller()` value to be used instead for otel trace "pkg" param and the `source_file` attribute